test/fixtures/rules/std_rule_cases/L050.yml

```diff
@@ -0,0 +1,24 @@
+rule: L050
+
+test_pass_leading_whitespace_statement:
+  pass_str: "SELECT foo FROM bar\n"
+
+test_pass_leading_whitespace_comment:
+  pass_str: "/*I am a comment*/\nSELECT foo FROM bar\n"
+
+test_pass_leading_whitespace_inline_comment:
+  pass_str: "--I am a comment\nSELECT foo FROM bar\n"
+
+test_pass_leading_whitespace_inline_comment_hash:
+  pass_str: "# I am a comment\nSELECT foo FROM bar\n"
+  configs:
+    core:
+      dialect: bigquery
+
+test_fail_leading_whitespace_statement:
+  fail_str: "\n  SELECT foo FROM bar\n"
+  fix_str: "SELECT foo FROM bar\n"
+
+test_fail_leading_whitespace_comment:
+  fail_str: "\n  /*I am a comment*/\nSELECT foo FROM bar\n"
+  fix_str: "/*I am a comment*/\nSELECT foo FROM bar\n"
```

Comment:
Given the problems with DBT (and seems those test cases cover this), I wonder should be add a `Jinja` test case.

```suggestion
  fix_str: "/*I am a comment*/\nSELECT foo FROM bar\n"
test_fail_leading_whitespace_jinja_comment:
  pass_str: "{# I am a comment #}\nSELECT foo FROM bar\n"
```

Reply:
@tunetheweb @barrywhart I've added a jinja pass test.

With regards to a query like this:
```sql

{# I am a comment #}
SELECT foo FROM bar
```
It has leading whitespace and the template block is detected (this one is included in the parse tree) and given fixes to delete the preceding whitespace by rule L050 however these fixes then get discarded and not raised at the top level. 
I believe this happens in `BaseRule.discard_unsafe_fixes` to not take the risk of overlapping templates, so as far as I'm aware the fact that this specific example is unchanged is an intended effect?

Reply:
But would `"\n{# I am a comment #}\nSELECT foo FROM bar\n"` not fail? It's just it wouldn't have a `fix_str` as it's "unfixable" automatically.

So the fact the above test passes is still a good test?

I presume `pass_str` actually tests it passes and doesn't just do a "before fix" and "after fix" comparison so "passes" even if it's not passed?

Reply:
@tunetheweb Good news, I understand why `"\n{# I am a comment #}\nSELECT foo FROM bar\n"` is passing rule 50 and it's intentional behaviour.
Within `linter.py` there is a function that filters out all fixes to templated code section unless it is a rule that `targets_templates` (just L046):
![image](https://user-images.githubusercontent.com/80432516/140802186-2a8c6c9a-082d-4f79-8529-22c9140a0d68.png)
OR if the user specifies that they want to lint templated areas (defaults to ignore them)
![image](https://user-images.githubusercontent.com/80432516/140802305-e4c21967-d69e-4dd7-914b-1a8287664fe6.png)

Reply:
So now if I change ignore_templated_areas to false in .sqlfluff, `"\n{# I am a comment #}\nSELECT foo FROM bar\n"` can be linted/fixed :)

Reply:
![image](https://user-images.githubusercontent.com/80432516/140803471-2305071f-67c9-4e17-8594-b2428bca11a6.png)

Reply:
`ignore_templated_areas` is not configurable. It's set per rule in the code.

This one should probably be set to true.

Reply:
Oh nevermind these yaml tests aren't testing the final outputed violations (post templated fix removals) but instead are just looking to see whether or not the violation is raised at any point in the lint, which it rightly is.
![image](https://user-images.githubusercontent.com/80432516/140806363-86307fc5-dcb6-4a39-a53d-40d6d1f8f9da.png)

Guess I can also add a custom test to show the impact of config on this rule

Reply:
so the pass version cares about the log not the output, but the fail version cares about the output not the log
![image](https://user-images.githubusercontent.com/80432516/140807519-b798b7cf-3dbb-4b41-b423-df06d53bc8c6.png)

Reply:
> `ignore_templated_areas` is not configurable. It's set per rule in the code.
> 
> This one should probably be set to true.

You're thinking of `targets_templated`, I believe it's specifically for template related rules i.e. L046. This rule should work if I were to set that attribute to true but feel thats not the intention of the attribute.

`ignore_templated_areas` 100% comes from the config (see below). I've tested configuring it with .sqlfluff and can confirm that it is what causes L050 to be applied or not in the case of a leading template. I agree with the existing default config to leave it off by default and allow users to turn it on if they wish.
![image](https://user-images.githubusercontent.com/80432516/140816093-a621a388-61d1-4294-8088-9e71ba7e454b.png)

Reply:
Ah yes you’re right. Got the two confused.

At the moment what can be passed though the simple API is relatively limited (hence why I just added `exclude_rules`). I’m happy to leave this as a positive test and not a negative test for now.

Reply:
Thanks for adding that `exclude_rules argument`, I'll account for that in my branch now

Reply:
> @tunetheweb yeah I was thinking we should really be able to pass a config object to the simple api too for easier testing, happy to pick that up as a separate PR :)

Actually I disagree. See reasoning here: https://github.com/sqlfluff/sqlfluff/pull/1419

Reply:
Makes sense to me, sound reasoning.

I've accounted for the exclude_rules addition in std_L003_L036_L039_combo_test.py 👍